### ISSUE TITLE
CA1812 false-positive when type is used as a generic type

### ISSUE THREAD

**Issue by meziantou** [dotnet/roslyn-analyzers#4025]
### Describe the bug

CA1812 false-positive

### Steps To Reproduce

````c#
    internal static class Repro
    {
        public static void Generate()
        {
            string str = "";
            var emojis = System.Text.Json.JsonSerializer.Deserialize<Dictionary<string, Emoji>>(str);
        }

        private sealed class Emoji // CA1812 whereas the type is used as a generic type
        {
            public string Name { get; set; }
            public string Moji { get; set; }
            public string Shortname { get; set; }
        }
    }
````

````xml
  <ItemGroup>
    <PackageReference Include="Microsoft.CodeAnalysis.FxCopAnalyzers" Version="3.3.0">
      <PrivateAssets>all</PrivateAssets>
      <IncludeAssets>runtime; build; native; contentfiles; analyzers; buildtransitive</IncludeAssets>
    </PackageReference>
  </ItemGroup>
````

### Expected behavior

No diagnostic

### Actual behavior

It reports a diagnostic that is not appliable in the context

### Analyzer

**Package**: [Microsoft.CodeAnalysis.FxCopAnalyzers](https://www.nuget.org/packages/Microsoft.CodeAnalysis.FxCopAnalyzers)

**Version**: v3.3.0 (Latest)

**Diagnostic ID**: [CA1812](https://docs.microsoft.com/visualstudio/code-quality/ca1812)

**Comment by mavasani** [comment]
This is same concern mentioned at https://github.com/dotnet/roslyn/issues/44330#issuecomment-630395417. As mentioned in https://github.com/dotnet/roslyn/issues/44330#issuecomment-631487553 the analyzer should not handle this but user should use a source suppression to document this.

**Comment by mavasani** [comment]
Oh actually I think @Evangelink might have already fixed this?

**Comment by Evangelink** [comment]
@mavasani I can reproduce the issue but I am not entirely sure how to fix it. The first thing that comes to my mind is to say that if a type is used as a generic type then we don't report on it but that's killing too many valid cases. The other option would be to have a list of methods (maybe configurable) that are considered as creating instance of types.

**Comment by mavasani** [comment]
3 options here:
1. By design: User should add source suppression for each type used as generic argument to such special methods
2. Conservative: Do not report for any type used as a generic type argument anywhere. This seems to have too many false positives, so lets not do this.
3. Configuration: Add an editorconfig option to specify special methods or types whose generic type arguments are instantiated, and should be excluded from analysis.

I think 1. or 3. would be fine.

**Comment by mavasani** [comment]
@genlu suggested a potential 4th option: Use a special attribute that can be applied to type/method to mark this semantics.

**Comment by meziantou** [comment]
Would it make sense to use the `DynamicallyAccessedMembers` attribute? Most methods from the framework are already annotated. For instance, `Deserialize` is annotated with `PublicProperties | PublicFields | PublicConstructors` ([source](https://github.com/dotnet/runtime/blob/c1a21a031925ce6261fadcfed26e098ab1bf46a6/src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializer.Read.Stream.cs#L38)). This means the method looks for the constructors of T. This doesn't guarantee to call it, but this is a good indication it will try to create an instance of the object.

That's being said, I'm not sure this will help for the reported code as the `T` parameter is actually a `Dictionary<string, CustomType>`, so the attribute doesn't explain what to do with `CustomType`. But it will help for `Deserialize<CustomType>(...)`.

**Comment by geeknoid** [comment]
> @mavasani I can reproduce the issue but I am not entirely sure how to fix it. The first thing that comes to my mind is to say that if a type is used as a generic type then we don't report on it but that's killing too many valid cases. The other option would be to have a list of methods (maybe configurable) that are considered as creating instance of types.

Hi there. I'm curious why not systematically suppressing any cases where the given type is used as a generic is not the right answer? The rule is really trying to say "hey, you've got a type here that nobody is using". When it's used a generic, the type is in fact being used, so no diagnostic should be produced.

**Comment by Youssef1313** [comment]
> The other option would be to have a list of methods (maybe configurable) that are considered as creating instance of types.

If we go with this approach, we should consider `AddHostedService` as one of the methods in the list. See https://github.com/dotnet/roslyn-analyzers/issues/6086

**Comment by znakeeye** [comment]
What if you put a `new()` constraint on `JsonSerializer.Deserialize<T>()`? Or would that break?

Otherwise, a workaround is to have a helper method with the constraint:
```
public static class MySerializer
{
    public static TValue? Deserialize<TValue>(string json, JsonSerializerOptions? options = null)
        where TValue : new()
    {
        return JsonSerializer.Deserialize<TValue>(json, options);
    }
}
```

Obviously, this will only work for types with public parameterless constructors.

**Comment by mavasani** [comment]
I am going to move this issue to dotnet/runtime for them to help us with the best option. Let me list down all the options we have identified in this issue:

**Issue:** [CA1812](https://learn.microsoft.com/dotnet/fundamentals/code-analysis/quality-rules/ca1812?view=vs-2022) false positives from non-explicit type instantiations

**Potential solutions:**
1. By design: User should add source suppression for each type used as generic argument to such special methods
2. Conservative: Do not report CA1812 for any type used as a generic type argument anywhere. This can have many false negatives.
3. Options based Configuration: Add an editorconfig option to specify special methods or types whose generic type arguments are instantiated, and should be excluded from analysis.
4. Attribute based Annotation:
   1. Special attribute that can be added to generic methods to indicate it may instantiate types of its type arguments
   2. Use `DynamicallyAccessedMembers` attribute as explained in https://github.com/dotnet/roslyn-analyzers/issues/4025#issuecomment-786362621
5. DiagnosticSuppressors based suppression: https://github.com/dotnet/roslyn-analyzers/issues/3545 suggests that libraries that have methods which instantiate generic types should write their own DiagnosticSuppressor to suppress CA1812 warnings. This may be tricky as the suppressor would have to re-analyze the whole project to look for calls to such special methods in the compilation. Also expecting each library to write such a suppressor seems unscalable.
